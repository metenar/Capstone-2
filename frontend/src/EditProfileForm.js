import React,{useContext} from "react";
import CurrentUserContext from "./CurrentUserContext"
import 'bootstrap/dist/css/bootstrap.css';
import { useFormik } from 'formik';
import { Card, Button, CardBody, 
        Form,FormGroup,Label ,Input} from 'reactstrap';

const validate = values => {
    const errors = {}; 
    if (!values.password) {
        errors.password = 'Required';
    } else if (values.password.length < 6) {
        errors.password = 'Must be at least 6 characters';
    }

    if (!values.first_name) {
        errors.first_name = 'Required';
    } else if (values.first_name.length > 15) {
        errors.first_name = 'Must be 15 characters or less';
    }

    if (!values.last_name) {
        errors.last_name = 'Required';
    } else if (values.last_name.length > 20) {
        errors.last_name = 'Must be 20 characters or less';
    }
    
    if (!values.email) {
        errors.email = 'Required';
    } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
        errors.email = 'Invalid email address';
    }

    return errors;
  };


const EditProfileForm=({save})=>{
    const {currentUser}=useContext(CurrentUserContext)
    console.log(currentUser)
    const formik = useFormik({
        initialValues : {
            username  : currentUser.username,
            first_name : currentUser.firstName,
            last_name  : currentUser.lastName,
            email     : currentUser.email,
            password  : '',
            image : currentUser.image
        },
        validate, 
        onSubmit      : async (values,actions) => {
          try {
            let res=await save(currentUser.username,values);
            if(res.success){
            actions.setStatus({ message: 'Profile Updated' });
            }
          } catch (err) {
            formik.errors.password = err;
          }
        }
      });
    
    return (
        <div className="col-md-4 m-auto">
            <h2>Profile</h2>
            <Card className="card">
                <CardBody>
                    <Form onSubmit={formik.handleSubmit}>
                        <FormGroup>
                            <Label >Username</Label>
                            <p className="font-weight-bold">{formik.values.username}</p>
                            <img className="card-img rounded-circle user-img mt-2 ml-2 p-2" style={{width:'100px'}} src={formik.values.image} alt="user avatar"/>
                        </FormGroup>
                        <FormGroup>
                            <Label>First Name</Label>
                            <Input
                                type="text"
                                name="first_name"
                                value={formik.values.first_name}
                                onChange={formik.handleChange}
                                onBlur={formik.handleBlur}
                            />
                            {formik.touched.first_name && formik.errors.first_name ? <div className="alert alert-danger">{formik.errors.first_name}</div>:null}
                        </FormGroup>
                        <FormGroup>
                            <Label>Last Name</Label>
                            <Input
                                type="text"
                                name="last_name"
                                value={formik.values.last_name}
                                onChange={formik.handleChange}
                                onBlur={formik.handleBlur}
                            />
                            {formik.touched.last_name && formik.errors.last_name ? <div className="alert alert-danger">{formik.errors.last_name}</div>:null}
                        </FormGroup>
                        <FormGroup>
                            <Label>Email</Label>
                            <Input
                                type="email"
                                name="email"
                                value={formik.values.email}
                                onChange={formik.handleChange}
                                onBlur={formik.handleBlur}
                            />
                            {formik.touched.email && formik.errors.email ? <div className="alert alert-danger">{formik.errors.email}</div>:null}
                        </FormGroup>
                        <FormGroup>
                            <Label>Confirm password to make changes:</Label>
                            <Input
                                type="password"
                                name="password"
                                value={formik.values.password}
                                onChange={formik.handleChange}
                                onBlur={formik.handleBlur}
                            />
                            {formik.touched.password && formik.errors.password ? <div className="alert alert-danger">{formik.errors.password}</div>:null}
                        </FormGroup>
                        <FormGroup>
                            <Label for="image">Avatar image </Label>
                            <Input 
                                type="text"
                                id="image"
                                name="image"
                                value={formik.values.image}
                                placeholder="Enter your image"
                                onChange={formik.handleChange}
                                onBlur={formik.handleBlur}
                            />
                            {formik.touched.image && formik.errors.image ? <div className="alert alert-danger">{formik.errors.image}</div>:null}
                        </FormGroup>
                        <div className="text-success">
                        {formik.status && formik.status.message ? formik.status.message:""}
                        </div>
                        <Button type="submit" color="primary">Save Changes</Button>
                    </Form>
                </CardBody>
            </Card>
        </div>
    )
}
export default EditProfileForm;