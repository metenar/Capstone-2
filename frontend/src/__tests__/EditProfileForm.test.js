import { render,screen} from '@testing-library/react';
import { MemoryRouter, Route, Switch} from "react-router-dom";
import EditProfileForm from '../EditProfileForm';
import '@testing-library/jest-dom'
import CurrentUserContext from "../CurrentUserContext"

function renderUpdate(currentUser) {

    return render(
        <MemoryRouter initialEntries={["/profile"]}>
            <CurrentUserContext.Provider value={{currentUser}}>
                <Switch>
                    <Route exect path="/profile">
                        <EditProfileForm save={"mete"} />
                    </Route>
                </Switch>
            </CurrentUserContext.Provider>
        </MemoryRouter>
    );
  }

  const currentUser=[
    {
      username:"metenar",
      firstName:"Mete",
      lastName:"NAR",
      email:'metenar@gmail.com',
      password:"",
      image:"www.mete.jpg"
    }]

test('renders without crushing', () => {
  renderUpdate({currentUser});
});

test('should match the snapshot', () => {
  const {asFragment}=renderUpdate({currentUser});
  expect(asFragment()).toMatchSnapshot()
});

test('should Book Status visible', ()=>{
    renderUpdate({currentUser});
    expect(screen.queryAllByText('metenar')).not.toBeNull();
})