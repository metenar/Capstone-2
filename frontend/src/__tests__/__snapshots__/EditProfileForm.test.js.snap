// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="col-md-4 m-auto"
  >
    <h2>
      Profile
    </h2>
    <div
      class="card card"
    >
      <div
        class="card-body"
      >
        <form
          class=""
        >
          <div
            class="form-group"
          >
            <label
              class=""
            >
              Username
            </label>
            <p
              class="font-weight-bold"
            />
            <img
              alt="user avatar"
              class="card-img rounded-circle user-img mt-2 ml-2 p-2"
              style="width: 100px;"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class=""
            >
              First Name
            </label>
            <input
              class="form-control"
              name="first_name"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class=""
            >
              Last Name
            </label>
            <input
              class="form-control"
              name="last_name"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class=""
            >
              Email
            </label>
            <input
              class="form-control"
              name="email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class=""
            >
              Confirm password to make changes:
            </label>
            <input
              class="form-control"
              name="password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class=""
              for="image"
            >
              Avatar image 
            </label>
            <input
              class="form-control"
              id="image"
              name="image"
              placeholder="Enter your image"
              type="text"
              value=""
            />
          </div>
          <div
            class="text-success"
          />
          <button
            class="btn btn-primary"
            type="submit"
          >
            Save Changes
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
